components/tilt.py

```diff
@@ -1,43 +1,67 @@
 import math
 
-GEAR_RATIO: float
-ANGLE_ERROR_TOLERANCE: float
+from magicbot import feedback
+from rev import CANSparkMax
+from wpilib import DoubleSolenoid, DutyCycleEncoder, PneumaticsModuleType
+from wpimath.controller import ProfiledPIDControllerRadians
+from wpimath.trajectory import TrapezoidProfileRadians
+
+from ids import DioChannels, PhChannels, SparkMaxIds
+
 INTAKING_ANGLE: float = math.radians(90)
+ANGLE_ERROR_TOLERANCE: float = math.radians(1)
+MAX_ANGULAR_VELOCITY: float = 12.0
+MAX_ANGULAR_ACCELERATION: float = 0.5
+POSITIVE_SOFT_LIMIT_ANGLE: float = math.radians(90)
+NEGATIVE_SOFT_LIMIT_ANGLE: float = math.radians(-90)
+TILT_ENCODER_ANGLE_OFFSET: float = 0
 
 
 class Tilt:
-    # motor: Neo550 #This will need to change to the proper motor controller type
-    # encoder: This will likely be an absolute encoder like the one we used on the arm.
-    # ^^ note for this we want to use the absolute api from the start so we dont have a disaster at comp
-    # brake: solenoid #This will probably be a double solenoid
-
-    # goal_angle = 0.0
+    goal_angle = 0.0
 
     def __init__(self) -> None:
-        # Create the hardware object handles
+        self.motor = CANSparkMax(
+            SparkMaxIds.tilt_motor, CANSparkMax.MotorType.kBrushless
+        )
+        self.absolute_encoder = DutyCycleEncoder(DioChannels.tilt_absolute_encoder)
+        self.absolute_encoder.setDistancePerRotation(math.tau)
+        self.absolute_encoder.setPositionOffset(TILT_ENCODER_ANGLE_OFFSET)
 
-        pass
+        self.brake = DoubleSolenoid(
+            PneumaticsModuleType.REVPH,
+            PhChannels.tilt_brake_forward,
+            PhChannels.tilt_brake_reverse,
+        )
+
+        rotation_contraints = TrapezoidProfileRadians.Constraints(
+            maxVelocity=MAX_ANGULAR_VELOCITY, maxAcceleration=MAX_ANGULAR_ACCELERATION
+        )
+        self.rotation_controller = ProfiledPIDControllerRadians(
+            32.0, 0.0, 0.0, rotation_contraints
+        )
 
     def set_angle(self, angle: float) -> None:
         # set the desired angle for the turret
+        clamped_angle = min(angle, POSITIVE_SOFT_LIMIT_ANGLE)
+        clamped_angle = max(clamped_angle, NEGATIVE_SOFT_LIMIT_ANGLE)
+        self.desired_angle = clamped_angle
```

Comment:
This sets `desired_angle` but everywhere else uses `goal_angle` which is only ever set to `0.0`. I presume this is meant to be `goal_angle` too?

```suggestion
        self.goal_angle = clamped_angle
```